File: C++/043. Multiply Strings.cpp
class Solution {
public:
    string multiply(string num1, string num2) 
    {
        int m = num1.length();
        int n = num2.length();
        if (m * n == 0) return "";
        if (m == 1 && num1[0] == '0') return "0";
        if (n == 1 && num2[0] == '0') return "0";
        int carry = 0;
        string res(m + n, '0');
        for (int k = m + n - 1; k >= 0; k--)
        {
            int val = carry;
            for (int i = max(0, k - n); i <= min(m - 1, k - 1); i++)
            {
                val += (num1[i] - '0') * (num2[k - 1 -i] - '0');
            }
            res[k] = (val % 10) + '0';
            carry = val / 10;
        }
        if (res[0] == '0')
        {
            res = res.substr(1);
        }
        return res;
    }
};

File: python/043. Multiply Strings.py
class Solution(object):
    def multiply(self, num1, num2):
        """
        :type num1: str
        :type num2: str
        :rtype: str
        """
        m = len(num1)
        n = len(num2)
        if m * n == 0: return ""
        if num1 == "0" or num2 == "0": return "0"
        res = ""
        carry = 0
        for k in range(n + m - 1):
            num = 0
            for i in range(max(0,k-n+1), min(m-1,k)+1):
                j = k - i
                num += int(num1[m-1-i]) * int(num2[n-1-j])
            res += str((num + carry) % 10)
            carry = (num + carry) / 10
        if carry > 0: res += str(carry)
        return res[::-1]